### PR TITLE
feat: Generate a request id for the recorded query

### PR DESCRIPTION
**Pull request by lynnagara**
This will be stored in the querylog table in https://github.com/getsentry/snuba/pull/783

### CODE REVIEW

**Review by tkaemming — CHANGES_REQUESTED** [review]
I think this is fine, but it might be better to add the ID to the `Request` when it is instantiated, in case there is anything else that could use it as a correlation ID. That field would then already be accessible via `SnubaQueryMetadata.request.id`. I don't have any specific ideas where that might be used right now, but it seems like it could be useful later and is not detrimental to the overall design. Calling it `request_id` would then also make it less likely to be confused with the ClickHouse `query_id` which would have a significantly different behavior.

**Comment by lynnagara** [comment]
Moved into the request

File: snuba/web/query_metadata.py [hunk 1]
```diff
@@ -35,6 +35,7 @@ class SnubaQueryMetadata:
 
     def to_dict(self) -> Mapping[str, Any]:
         return {
+            "query_id": self.request.id,
```
Comment (tkaemming) [inline snuba/web/query_metadata.py]:
```suggestion
            "request_id": self.request.id,
```

**Review by tkaemming — APPROVED** [review]
This looks better, but is there a reason that `SnubaQueryMetadata` includes only the request body and not the other attributes?

**Comment by lynnagara** [comment]
Mostly because that's all we are planning to store in the querylog right now, is there's something else you'd like to see in there?

**Comment by tkaemming** [comment]
> Mostly because that's all we are planning to store in the querylog table right now, is there's something else you'd like to see in there?

It seems to me that it'd make sense for the request attributes in this data structure (id, referrer, body) all be to part of the same data structure instead of merged into top level attributes with other data. `Request.body` is also deprecated in favor of more clearly defined attributes defined in the class definition, at least according to this comment: https://github.com/getsentry/snuba/blob/9ee56eac9e1edd63decca4d1fd56002617ff949f/snuba/request/__init__.py#L20-L23

**Comment by lynnagara** [comment]
Why do you want to put the nested structure into redis instead of the flattened out one? I'm going to merge this as-is and replace request.body with something custom that converts the attributes into a string in the other PR.

**Comment by fpacifici** [comment]
Just for context, the query body there represent the query provided by the user before any processing (well, not true, the extensions are already applied, but that's as good as it gets at this point of query processing, we should try to get the original query body from the request).
So it was deprecated for query processing, if we need it to store in stats for observability we could revisit that and maybe make it immutable.

**Comment by lynnagara** [comment]
Yeah, ideally I wanted to get the original request here, this is just the closest we have

**Comment by tkaemming** [comment]
> Why do you want to put the nested structure into redis instead of the flattened out one?

I'm not sure if you're talking about the `Request` instance in general or just the body, but regardless my opinion is that the things that we log should reflect the actual data model as closely and consistently as possible so they are easy to understand and work with, without being concerned about the implementation differences between them.

If a `Request` is represented one way during the actual request/response cycle and in a different form in the query log, it can be confusing to remember which flavor of request you are working with in different contexts. If we add more attributes to the `Request`, it also means that code for the query log itself has to change. (This can be a positive or a benefit as well, but I think it is generally more good than bad.)

Ideally, you'd be able to just pull the `Request` object out of the query log and pass it to `parse_and_run_query` without having to reconstitute it from its component parts. (This is invaluable if you need to rerun the query for some reason, we did this a lot when developing Snuba.) That's not possible as implemented here.

> Just for context, the query body there represent the query provided by the user before any processing (well, not true, the extensions are already applied, but that's as good as it gets at this point of query processing, we should try to get the original query body from the request). So it was deprecated for query processing, if we need it to store in stats for observability we could revisit that and maybe make it immutable.

That's fine, but I don't think we're/you're actually looking for the (Snuba) `Request` body here, then, and instead the _HTTP_ request body should be logged. The parenthetical is even evidence why it's confusing to treat them as the same object.

This conflates them in a way that we end up without an accurate representation of the Snuba or HTTP request. I think it'd be okay to log the HTTP request body, with the exception that it wouldn't work for subscription requests, and some detail is lost (anything in the URL or headers, for instance.)

**Pull request merged**